### PR TITLE
Golang: Noise confidence interval for Gaussian noise

### PR DESCRIPTION
**Pull request by laviniamnedelea**
- Added Confidence interval computation for Gaussian noise
- Added tests for inverseCDFGaussian and confidence interval interface
  functions

### CODE REVIEW

File: go/noise/gaussian_noise.go [hunk 1]
```diff
@@ -155,6 +186,23 @@ func addGaussian(x, sigma float64) float64 {
 	return roundToMultipleOfPowerOfTwo(x, granularity) + float64(sample)*granularity
 }
 
+// getConfidenceIntervalGaussian computes the confidence interval using the inverse error function.
+func getConfidenceIntervalGaussian(noisedValue, confidenceLevel, sigma float64) ConfidenceInterval {
+	shiftingValue := inverseCDFGaussian(sigma, confidenceLevel)
```
Comment (dkrft) [inline go/noise/gaussian_noise.go:191]: inverseCDFGaussian should be either called on (sigma, 1 - alpha / 2) or (sigma, alpha / 2). in the latter case (which is what I would suggest to do), you need to add the result noisedX for the lower bound and subtract it from the upperBound (because the result will be negative).

Comment (dkrft) [inline go/noise/gaussian_noise.go:191]: I am not sure I like the name "shiftingValue". I also don't have a good alternative suggestion. Maybe something short like "z" works here. I think it makes sense to be conistent with the Lapalce implemntation here.

Comment (laviniamnedelea) [reply go/noise/gaussian_noise.go:191]: That makes sense, thank you. I will switch it to z. :)

File: go/go.sum [hunk 2]
```diff
@@ -0,0 +1,23 @@
+github.com/ajstarks/svgo v0.0.0-20180226025133-644b8db467af/go.mod h1:K08gAheRH3/J6wwsYMMT4xOr94bZjxIelGM0+d/wbFw=
+github.com/fogleman/gg v1.2.1-0.20190220221249-0403632d5b90/go.mod h1:R/bRT+9gY/C5z7JzPU0zXsXHKM4/ayA+zqcVNZzPa1k=
```
Comment (MashaTelyatnikova) [inline go/go.sum:2]: can you remove this file from the PR?

File: go/noise/gaussian_noise.go [hunk 3]
```diff
@@ -155,6 +181,23 @@ func addGaussian(x, sigma float64) float64 {
 	return roundToMultipleOfPowerOfTwo(x, granularity) + float64(sample)*granularity
 }
 
+// computeConfidenceIntervalGaussian computes a confidence interval that contains the raw value x from which
+// float64 noisedX is computed with a probability equal to 1 - alpha with the given sigma
+func computeConfidenceIntervalGaussian(noisedX, sigma, alpha float64) ConfidenceInterval {
+	z := inverseCDFGaussian(sigma, alpha/2)
```
Comment (MashaTelyatnikova) [inline go/noise/gaussian_noise.go:187]: why z? Can the name be more meaningful?

File: go/noise/gaussian_noise.go [hunk 4]
```diff
@@ -155,6 +181,23 @@ func addGaussian(x, sigma float64) float64 {
 	return roundToMultipleOfPowerOfTwo(x, granularity) + float64(sample)*granularity
 }
 
+// computeConfidenceIntervalGaussian computes a confidence interval that contains the raw value x from which
+// float64 noisedX is computed with a probability equal to 1 - alpha with the given sigma
```
Comment (MashaTelyatnikova) [inline go/noise/gaussian_noise.go:185]: add a full-stop at the and of all comments which describe methods

File: go/noise/gaussian_noise.go [hunk 5]
```diff
@@ -120,16 +121,33 @@ func (gaussian) DeltaForThreshold(l0Sensitivity int64, lInfSensitivity, epsilon,
...
+	return computeConfidenceIntervalGaussian(float64(noisedX), sigma, alpha).roundToInt64(), nil
 }
 
-func (gaussian) computeConfidenceIntervalFloat64(noisedValue float64, l0Sensitivity int64, lInfSensitivity, epsilon, delta, confidenceLevel float64) (ConfidenceInterval, error) {
-	// TODO: Implement confidence interval computation.
-	return ConfidenceInterval{}, nil
+// computeConfidenceIntervalFloat64 computes a confidence interval that contains the raw value x from which float64
+// noisedX is computed with a probability equal to 1 - alpha based on the specified gaussian noise parameters.
+func (gaussian) computeConfidenceIntervalFloat64(noisedX float64, l0Sensitivity int64, lInfSensitivity, epsilon, delta, alpha float64) (ConfidenceInterval, error) {
+	if err := checkArgsConfidenceIntervalGaussian("GaussianNoiseInterval", l0Sensitivity, lInfSensitivity, epsilon, delta,
```
Comment (MashaTelyatnikova) [inline go/noise/gaussian_noise.go:140]: label can be "ComputeConfidenceIntervalFloat64"
the purpose of labels to identify (in the log) which method caused the error.
for the int64 put "ComputeConfidenceIntervalInt64"

File: go/noise/gaussian_noise.go (hunk 3 above)
Comment (laviniamnedelea) [reply go/noise/gaussian_noise.go:187]: At first i chose shiftingValue for the name, but Dennis suggested something shorter and consistent with the Laplace implementation. Neither me, nor him, had any better ideas for the name at the moment.

File: go/noise/noise.go [hunk 6]
```diff
@@ -114,13 +114,11 @@ type Noise interface {
 	// given assumptions of L_0 and L_∞ sensitivities.
 	Threshold(l0Sensitivity int64, lInfSensitivity, epsilon, noiseDelta, thresholdDelta float64) float64
 
-	// confidenceIntervalInt64 will return a ConfidenceInterval struct with the given
-	// confidenceLevel, using the int64 noisedValue, l0Sensitivity, lInfSensitivity int64 and
-	// epsilon, delta float64 for the distribution.
-	computeConfidenceIntervalInt64(noisedValue, l0Sensitivity, lInfSensitivity int64, epsilon, delta, confidenceLevel float64) (ConfidenceInterval, error)
+	// computeConfidenceIntervalInt64 computes a confidence interval that contains the raw integer value x from which int64
+	// noisedX is computed with a probability greater or equal to 1 - alpha based on the specified noise parameters.
+	computeConfidenceIntervalInt64(noisedX, l0Sensitivity, lInfSensitivity int64, epsilon, delta, alpha float64) (ConfidenceInterval, error)
```
Comment (miracvbasaran) [inline go/noise/noise.go:119]: Please make this exported.

File: go/noise/noise.go [hunk 7]
```diff
@@ -114,13 +114,11 @@ type Noise interface {
...
 
-	// confidenceIntervalFloat64 will return a ConfidenceInterval struct with the given
-	// confidenceLevel, using the float64 noisedValue, l0Sensitivity int64 and lInfSensitivity,
-	// epsilon, delta float64 for the distribution.
-	computeConfidenceIntervalFloat64(noisedValue float64, l0Sensitivity int64, lInfSensitivity, epsilon, delta, confidenceLevel float64) (ConfidenceInterval, error)
+	// computeConfidenceIntervalFloat64 computes a confidence interval that contains the raw value x from which float64
+	// noisedX is computed with a probability equal to 1 - alpha based on the specified noise parameters.
+	computeConfidenceIntervalFloat64(noisedX float64, l0Sensitivity int64, lInfSensitivity, epsilon, delta, alpha float64) (ConfidenceInterval, error)
```
Comment (miracvbasaran) [inline go/noise/noise.go:123]: Please make this exported.

File: go/noise/laplace_noise.go [hunk 8]
```diff
@@ -164,12 +164,12 @@ func (laplace) DeltaForThreshold(l0Sensitivity int64, lInfSensitivity, epsilon,
 	return 1 - math.Pow(1-partitionDelta, float64(l0Sensitivity))
 }
 
-func (laplace) computeConfidenceIntervalInt64(noisedValue, l0Sensitivity, lInfSensitivity int64, epsilon, delta, confidenceLevel float64) (ConfidenceInterval, error) {
+func (laplace) computeConfidenceIntervalInt64(noisedX, l0Sensitivity, lInfSensitivity int64, epsilon, delta, alpha float64) (ConfidenceInterval, error) {
```
Comment (miracvbasaran) [inline go/noise/laplace_noise.go:167]: Please make this exported.

File: go/noise/laplace_noise.go [hunk 9]
```diff
@@ -164,12 +164,12 @@ func (laplace) DeltaForThreshold(l0Sensitivity int64, lInfSensitivity, epsilon,
...
 	// TODO: Implement confidence interval computation.
 	return ConfidenceInterval{}, nil
 }
 
-func (laplace) computeConfidenceIntervalFloat64(noisedValue float64, l0Sensitivity int64, lInfSensitivity, epsilon, delta, confidenceLevel float64) (ConfidenceInterval, error) {
+func (laplace) computeConfidenceIntervalFloat64(noisedX float64, l0Sensitivity int64, lInfSensitivity, epsilon, delta, alpha float64) (ConfidenceInterval, error) {
```
Comment (miracvbasaran) [inline go/noise/laplace_noise.go:172]: Please make this exported.

File: go/noise/gaussian_noise.go [hunk 10]
```diff
@@ -155,6 +181,23 @@ func addGaussian(x, sigma float64) float64 {
 	return roundToMultipleOfPowerOfTwo(x, granularity) + float64(sample)*granularity
 }
 
+// computeConfidenceIntervalGaussian computes a confidence interval that contains the raw value x from which
+// float64 noisedX is computed with a probability equal to 1 - alpha with the given sigma.
+func computeConfidenceIntervalGaussian(noisedX, sigma, alpha float64) ConfidenceInterval {
+	z := inverseCDFGaussian(sigma, alpha/2)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise.go:187]: Maybe comment that this is negative? It might be confusing since we say lowerbound is `noisedX + z`

File: go/noise/gaussian_noise.go [hunk 11]
```diff
@@ -155,6 +181,23 @@ func addGaussian(x, sigma float64) float64 {
...
+	FloatLowerBound := noisedX + z
+	FloatUpperBound := noisedX - z
+	FloatInterval := ConfidenceInterval{LowerBound: FloatLowerBound,
+		UpperBound: FloatUpperBound}
+	return FloatInterval
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise.go:192]: Please make this a single line.

File: go/noise/gaussian_noise_test.go [hunk 12]
```diff
@@ -356,70 +356,70 @@ var thresholdGaussianTestCases = []struct {
 	l0Sensitivity   int64
 	lInfSensitivity float64
 	epsilon         float64
-	noiseDelta      float64
-	thresholdDelta  float64
+	deltaNoise      float64
+	deltaThreshold  float64
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:360]: Could you please redo these changes by rebasing to main?

File: go/noise/gaussian_noise_test.go [hunk 13]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+				t.Errorf("Got delta: %0.12f, want delta: %0.12f", gotDelta, tc.deltaThreshold)
 			}
 		})
 	}
 }
+
+func TestInverseCDFGaussian(t *testing.T) {
+	for _, tc := range []struct {
+		desc           string
+		sigma, p, want float64 // Where p is equal to alpha/2.
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:444]: Could you please make these multiple lines? That's the idiomatic way of doing things for test structs.

File: go/noise/gaussian_noise_test.go [hunk 14]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+			p:     10e-10,
+			want:  -2.1352192973427364,
+		},
+		{
+			desc:  "Edge case test with high alpha",
+			sigma: 0.84,
+			p:     1 - 10e-10,
+			want:  5.0381578964653757,
+		},
+		// For p = 0.5, the result should be the mean regardless of sigma
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:477]: Nit: `.` at the end.

File: go/noise/gaussian_noise_test.go [hunk 15]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+			want:  0,
+		},
+		{
+			desc:  "Test with p = 0.5",
+			sigma: 0.8235243,
+			p:     0.5,
+			want:  0,
+		},
+	} {
+
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:491]: Please remove the unnecessary newline here.

File: go/noise/gaussian_noise_test.go [hunk 16]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+			noisedX: 99.98989898,
+			sigma:   15423235,
+			alpha:   1 - 0.111,
+			want:    ConfidenceInterval{-2.1525159435946424e+06, 2.1527159233926027e+06},
+		},
+		{
+			desc:    "Low confidence level",
+			noisedX: 100,
+			sigma:   10,
+			alpha:   1 - 10e-10,
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:541]: Nit: `1e-10` instead of `10e-10`

File: go/noise/gaussian_noise_test.go [hunk 17]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+			want:    ConfidenceInterval{99.99999998746686, 100.00000001253314},
+		},
+		{
+			desc:    "High confidence level",
+			noisedX: 100,
+			sigma:   10,
+			alpha:   10e-10,
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:548]: Nit: `1e-10` instead of `10e-10`

File: go/noise/gaussian_noise_test.go [hunk 18]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+		result := computeConfidenceIntervalGaussian(tc.noisedX, tc.sigma, tc.alpha)
+		if !approxEqual(result.LowerBound, tc.want.LowerBound) {
+			t.Errorf("TestComputeConfidenceIntervalGaussian(%f, %f, %f)=%0.10f, want %0.10f, desc %s, LowerBound is not equal",
+				tc.noisedX, tc.alpha, tc.sigma,
+				result.LowerBound, tc.want.LowerBound, tc.desc)
+		}
+		if !approxEqual(result.UpperBound, tc.want.UpperBound) {
+			t.Errorf("TestConfidenceIntervalLaplace(%f, %f, %f)=%0.10f, want %0.10f, desc %s, UpperBound is not equal",
+				tc.noisedX, tc.alpha, tc.sigma,
+				result.UpperBound, tc.want.UpperBound, tc.desc)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:561]: Please make this single line.

File: go/noise/gaussian_noise_test.go [hunk 19]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+			sigma:   10,
+			alpha:   10e-10,
+			want:    ConfidenceInterval{38.90589790616554, 161.09410209383446},
+		},
+	} {
+		result := computeConfidenceIntervalGaussian(tc.noisedX, tc.sigma, tc.alpha)
+		if !approxEqual(result.LowerBound, tc.want.LowerBound) {
+			t.Errorf("TestComputeConfidenceIntervalGaussian(%f, %f, %f)=%0.10f, want %0.10f, desc %s, LowerBound is not equal",
+				tc.noisedX, tc.alpha, tc.sigma,
+				result.LowerBound, tc.want.LowerBound, tc.desc)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:556]: Please make this single line.

File: go/noise/gaussian_noise_test.go [hunk 20]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+		}
+	}
+
+}
+
+func TestComputeConfidenceIntervalInt64(t *testing.T) {
+	for _, tc := range []struct {
+		desc                                    string
+		noisedX, l0Sensitivity, lInfSensitivity int64
+		epsilon, delta, alpha                   float64
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:571]: Please make these multiple lines.

File: go/noise/gaussian_noise_test.go [hunk 21]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+			lInfSensitivity: 36,
+			epsilon:         0.8,
+			delta:           0.8,
+			alpha:           math.NaN(),
+			want:            ConfidenceInterval{},
+			wantErr:         true,
+		},
+	} {
+		got, err := gauss.computeConfidenceIntervalInt64(tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity,
+			tc.epsilon, tc.delta, tc.alpha)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:808]: Please make this single line.

File: go/noise/gaussian_noise_test.go [hunk 22]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+		if (err != nil) != tc.wantErr {
+			t.Errorf("computeConfidenceIntervalInt64: when %s for err got %v", tc.desc, err)
+			continue
+		}
+		if got.LowerBound != tc.want.LowerBound {
+			t.Errorf("TestComputeConfidenceIntervalInt64(%d, %d, %d, %f, %f, %f)=%f, want %f, desc %s, LowerBound is not equal",
+				tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity, tc.epsilon, tc.delta, tc.alpha,
+				got.LowerBound, tc.want.LowerBound, tc.desc)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:816]: Please make this single line.

File: go/noise/gaussian_noise_test.go [hunk 23]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+		}
+		if got.UpperBound != tc.want.UpperBound {
+			t.Errorf("TestComputeConfidenceIntervalInt64(%d, %d, %d, %f, %f, %f)=%f, want %f, desc %s, UpperBound is not equal",
+				tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity, tc.epsilon, tc.delta, tc.alpha,
+				got.UpperBound, tc.want.UpperBound, tc.desc)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:821]: Please make this single line.

File: go/noise/gaussian_noise_test.go [hunk 24]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+		}
+	}
+}
+
+func TestComputeConfidenceIntervalFloat644(t *testing.T) {
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:826]: Nit: `Float64`

File: go/noise/gaussian_noise_test.go [hunk 25]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+	for _, tc := range []struct {
+		desc                                   string
+		noisedX                                float64
+		l0Sensitivity                          int64
+		lInfSensitivity, epsilon, delta, alpha float64
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:831]: Please make these multiple lines.

File: go/noise/gaussian_noise_test.go [hunk 26]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+			lInfSensitivity: 5,
+			epsilon:         0.05,
+			delta:           math.Inf(1), // delta should be strictly positive and smaller than 1.
+			alpha:           0.2,
+			want:            ConfidenceInterval{},
+			wantErr:         true,
+		},
+	} {
+		got, err := gauss.computeConfidenceIntervalFloat64(tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity,
+			tc.epsilon, tc.delta, tc.alpha)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:1069]: Please make this single line.

File: go/noise/gaussian_noise_test.go [hunk 27]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+		if (err != nil) != tc.wantErr {
+			t.Errorf("computeConfidenceIntervalFloat644: when %s for err got %v", tc.desc, err)
+			continue
+		}
+		if !approxEqual(got.LowerBound, tc.want.LowerBound) {
+			t.Errorf("TestComputeConfidenceIntervalFloat644(%f, %d, %f, %f, %f)=%0.10f, want %0.10f, desc %s, LowerBound is not equal",
+				tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity, tc.epsilon, tc.alpha,
+				got.UpperBound, tc.want.UpperBound, tc.desc)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:1077]: Please make this single line.

File: go/noise/gaussian_noise_test.go [hunk 28]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+		}
+		if !approxEqual(got.UpperBound, tc.want.UpperBound) {
+			t.Errorf("TestComputeConfidenceIntervalFloat644(%f, %d, %f, %f, %f)=%0.10f, want %0.10f, desc %s, UpperBound is not equal",
+				tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity, tc.epsilon, tc.alpha,
+				got.LowerBound, tc.want.LowerBound, tc.desc)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:1082]: Please make this single line.

File: go/noise/gaussian_noise_test.go [hunk 29]
```diff
@@ -430,10 +430,656 @@ func TestThresholdGaussian(t *testing.T) {
...
+			alpha:           0.2,
+			want:            ConfidenceInterval{},
+			wantErr:         true,
+		},
+	} {
+		got, err := gauss.computeConfidenceIntervalFloat64(tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity,
+			tc.epsilon, tc.delta, tc.alpha)
+		if (err != nil) != tc.wantErr {
+			t.Errorf("computeConfidenceIntervalFloat644: when %s for err got %v", tc.desc, err)
+			continue
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:1072]: Does this `continue` do anything? If not, we should remove it. (Also above)

File: go/noise/gaussian_noise.go [hunk 30]
```diff
@@ -143,6 +161,14 @@ func checkArgsGaussian(label string, l0Sensitivity int64, lInfSensitivity, epsil
...
 }
 
+// checkArgsConfidenceIntervalGaussian checks the parameters for gaussian confidence interval, as well as the provided confidence level.
+func checkArgsConfidenceIntervalGaussian(label string, l0Sensitivity int64, lInfSensitivity, epsilon, delta, alpha float64) error {
+	if err := checks.CheckAlpha(label, alpha); err != nil {
+		return err
+	}
+	return checkArgsGaussian(label, l0Sensitivity, lInfSensitivity, epsilon, delta)
+}
+
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise.go:171]: I believe this should live in checks.go

File: go/noise/gaussian_noise.go [hunk 31]
```diff
@@ -120,16 +121,33 @@ func (gaussian) DeltaForThreshold(l0Sensitivity int64, lInfSensitivity, epsilon,
...
 }
 
-func (gaussian) computeConfidenceIntervalFloat64(noisedValue float64, l0Sensitivity int64, lInfSensitivity, epsilon, delta, confidenceLevel float64) (ConfidenceInterval, error) {
-	// TODO: Implement confidence interval computation.
-	return ConfidenceInterval{}, nil
+// computeConfidenceIntervalFloat64 computes a confidence interval that contains the raw value x from which float64
+// noisedX is computed with a probability equal to 1 - alpha based on the specified gaussian noise parameters.
+func (gaussian) computeConfidenceIntervalFloat64(noisedX float64, l0Sensitivity int64, lInfSensitivity, epsilon, delta, alpha float64) (ConfidenceInterval, error) {
+	if err := checkArgsConfidenceIntervalGaussian("computeConfidenceIntervalFloat64", l0Sensitivity, lInfSensitivity, epsilon, delta,
+		alpha); err != nil {
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise.go:141]: Could you please make this single line?

File: go/noise/gaussian_noise.go [hunk 32]
```diff
@@ -120,16 +121,33 @@ func (gaussian) DeltaForThreshold(l0Sensitivity int64, lInfSensitivity, epsilon,
...
 }
 
-func (gaussian) computeConfidenceIntervalInt64(noisedValue, l0Sensitivity, lInfSensitivity int64, epsilon, delta, confidenceLevel float64) (ConfidenceInterval, error) {
-	// TODO: Implement confidence interval computation.
-	return ConfidenceInterval{}, nil
+// computeConfidenceIntervalInt64 computes a confidence interval that contains the raw integer value x from which int64 noisedX
+// is computed with a probability greater or equal to 1 - alpha based on the specified gaussian noise parameters.
+func (gaussian) computeConfidenceIntervalInt64(noisedX, l0Sensitivity, lInfSensitivity int64, epsilon, delta, alpha float64) (ConfidenceInterval, error) {
+	if err := checkArgsConfidenceIntervalGaussian("computeConfidenceIntervalInt64", l0Sensitivity, float64(lInfSensitivity), epsilon, delta,
+		alpha); err != nil {
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise.go:128]: Could you please make this single line?

File: go/noise/gaussian_noise.go [hunk 33]
```diff
@@ -155,6 +181,23 @@ func addGaussian(x, sigma float64) float64 {
...
+	FloatUpperBound := noisedX - z
+	FloatInterval := ConfidenceInterval{LowerBound: FloatLowerBound,
+		UpperBound: FloatUpperBound}
+	return FloatInterval
+}
+
+// inverseCDFGaussian computes the quantile z satisfying Pr[Y <= z] = p for a random variable Y that is Gaussian
+// distributed with the specified sigma.
+func inverseCDFGaussian(sigma, p float64) float64 {
+	return -sigma * math.Sqrt(2) * math.Erfcinv(2.00*p)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise.go:198]: Does this need to be `2.00`? Can we not use `2`?

File: go/noise/gaussian_noise.go [hunk 34]
```diff
@@ -155,6 +181,23 @@ func addGaussian(x, sigma float64) float64 {
...
+	z := inverseCDFGaussian(sigma, alpha/2)
+	FloatLowerBound := noisedX + z
+	FloatUpperBound := noisedX - z
+	FloatInterval := ConfidenceInterval{LowerBound: FloatLowerBound,
+		UpperBound: FloatUpperBound}
+	return FloatInterval
+}
+
+// inverseCDFGaussian computes the quantile z satisfying Pr[Y <= z] = p for a random variable Y that is Gaussian
+// distributed with the specified sigma.
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise.go:196]: Nit: `specified sigma and mean 0.`

**Review by miracvbasaran — CHANGES_REQUESTED** [review]

File: go/noise/gaussian_noise.go (hunk 30 above)
Comment (laviniamnedelea) [reply go/noise/gaussian_noise.go:171]: I also have the same comment Karim mentioned, so i will wait for your response in his PR :) Thank you!

File: go/noise/gaussian_noise.go (hunk 33 above)
Comment (laviniamnedelea) [reply go/noise/gaussian_noise.go:198]: It should be just fine with 2 as well :)

File: go/noise/gaussian_noise_test.go [hunk 35]
```diff
@@ -538,27 +538,25 @@ func TestComputeConfidenceIntervalGaussian(t *testing.T) {
...
 	} {
 		result := computeConfidenceIntervalGaussian(tc.noisedX, tc.sigma, tc.alpha)
 		if !approxEqual(result.LowerBound, tc.want.LowerBound) {
 			t.Errorf("TestComputeConfidenceIntervalGaussian(%f, %f, %f)=%0.10f, want %0.10f, desc %s, LowerBound is not equal",
-				tc.noisedX, tc.alpha, tc.sigma,
-				result.LowerBound, tc.want.LowerBound, tc.desc)
+				tc.noisedX, tc.alpha, tc.sigma, result.LowerBound, tc.want.LowerBound, tc.desc)
 		}
 		if !approxEqual(result.UpperBound, tc.want.UpperBound) {
 			t.Errorf("TestConfidenceIntervalLaplace(%f, %f, %f)=%0.10f, want %0.10f, desc %s, UpperBound is not equal",
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:558]: NIt: `Gaussian` instead of `Laplace`

File: go/noise/gaussian_noise_test.go [hunk 36]
```diff
@@ -804,26 +802,22 @@ func TestComputeConfidenceIntervalInt64(t *testing.T) {
 			wantErr:         true,
 		},
 	} {
-		got, err := gauss.computeConfidenceIntervalInt64(tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity,
-			tc.epsilon, tc.delta, tc.alpha)
+		got, err := gauss.ComputeConfidenceIntervalInt64(tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity, tc.epsilon, tc.delta, tc.alpha)
 		if (err != nil) != tc.wantErr {
-			t.Errorf("computeConfidenceIntervalInt64: when %s for err got %v", tc.desc, err)
-			continue
+			t.Errorf("ComputeConfidenceIntervalInt64: when %s for err got %v", tc.desc, err)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:807]: nit:`ComputeConfidenceIntervalInt64: when %s got err %v, wantErr=%t`

File: go/noise/gaussian_noise_test.go [hunk 37]
```diff
@@ -1065,21 +1059,17 @@ func TestComputeConfidenceIntervalFloat644(t *testing.T) {
 			wantErr:         true,
 		},
 	} {
-		got, err := gauss.computeConfidenceIntervalFloat64(tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity,
-			tc.epsilon, tc.delta, tc.alpha)
+		got, err := gauss.ComputeConfidenceIntervalFloat64(tc.noisedX, tc.l0Sensitivity, tc.lInfSensitivity, tc.epsilon, tc.delta, tc.alpha)
 		if (err != nil) != tc.wantErr {
-			t.Errorf("computeConfidenceIntervalFloat644: when %s for err got %v", tc.desc, err)
-			continue
+			t.Errorf("ComputeConfidenceIntervalFloat644: when %s for err got %v", tc.desc, err)
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise_test.go:1064]: nit:ComputeConfidenceIntervalFloat64: when %s got err %v, wantErr=%t

**Review by miracvbasaran — CHANGES_REQUESTED** [review]

File: go/noise/gaussian_noise.go [hunk 38]
```diff
@@ -184,18 +184,17 @@ func addGaussian(x, sigma float64) float64 {
 // computeConfidenceIntervalGaussian computes a confidence interval that contains the raw value x from which
 // float64 noisedX is computed with a probability equal to 1 - alpha with the given sigma.
 func computeConfidenceIntervalGaussian(noisedX, sigma, alpha float64) ConfidenceInterval {
-	z := inverseCDFGaussian(sigma, alpha/2)
+	z := inverseCDFGaussian(sigma, alpha/2) // z will hold a negative value.
 	FloatLowerBound := noisedX + z
 	FloatUpperBound := noisedX - z
-	FloatInterval := ConfidenceInterval{LowerBound: FloatLowerBound,
-		UpperBound: FloatUpperBound}
+	FloatInterval := ConfidenceInterval{LowerBound: FloatLowerBound, UpperBound: FloatUpperBound}
```
[the remark this answers is not in the document]
Comment (laviniamnedelea) [reply go/noise/gaussian_noise.go:190]: Ohh, i'm sorry, i misunderstood.

File: go/checks/checks_test.go [hunk 39]
```diff
@@ -528,7 +528,7 @@ func TestCheckAlpha(t *testing.T) {
 		{"arbitrary alpha", 0.758464984, false},
 	} {
 		if err := CheckAlpha("test", tc.alpha); (err != nil) != tc.wantErr {
-			t.Errorf("CheckAlpha: when %s for err got %v, want %t", tc.desc, err, tc.wantErr)
+			t.Errorf("checkAlpha: when %s for err got %v, want %t", tc.desc, err, tc.wantErr)
```
Comment (miracvbasaran) [inline go/checks/checks_test.go:531]: Please rename to `CheckAlpha` since that's the name of the function.

File: go/noise/gaussian_noise.go [hunk 40]
```diff
@@ -120,16 +121,33 @@ func (gaussian) DeltaForThreshold(l0Sensitivity int64, lInfSensitivity, epsilon,
...
+	if err != nil {
+		err = fmt.Errorf("ComputeConfidenceIntervalFloat64(l0sensitivity %d, lInfSensitivity %f, epsilon %f, delta %e, alpha %f) checks failed with %v",
+			l0Sensitivity, lInfSensitivity, epsilon, delta, alpha, err)
+		return ConfidenceInterval{}, err
+	}
+	sigma := SigmaForGaussian(l0Sensitivity, lInfSensitivity, epsilon, delta)
+	return computeConfidenceIntervalGaussian(noisedX, sigma, alpha), nil
 }
 
+// checkArgsGaussian checks the parameters for gaussian confidence interval.
```
Comment (miracvbasaran) [inline go/noise/gaussian_noise.go:150]: Please remove this since it's not true.

File: go/noise/laplace_noise.go [hunk 41]
```diff
@@ -165,30 +164,14 @@ func (laplace) DeltaForThreshold(l0Sensitivity int64, lInfSensitivity, epsilon,
...
 func (laplace) ComputeConfidenceIntervalInt64(noisedX, l0Sensitivity, lInfSensitivity int64, epsilon, delta, alpha float64) (ConfidenceInterval, error) {
-	err := checkArgsConfidenceIntervalLaplace("ComputeConfidenceIntervalInt64 (Laplace)", l0Sensitivity, float64(lInfSensitivity), epsilon, delta, alpha)
-	if err != nil {
-		err = fmt.Errorf("ComputeConfidenceIntervalInt64(noisedX %d, l0sensitivity %d, lInfSensitivity %d, epsilon %f, delta %e, alpha %f) checks failed with %v",
-			noisedX, l0Sensitivity, lInfSensitivity, epsilon, delta, alpha, err)
-		return ConfidenceInterval{}, err
-	}
-	lambda := laplaceLambda(l0Sensitivity, float64(lInfSensitivity), epsilon)
-	return computeConfidenceIntervalLaplace(float64(noisedX), lambda, alpha).roundToInt64(), nil
+	// TODO: Implement confidence interval computation.
```
Comment (miracvbasaran) [inline go/noise/laplace_noise.go:168]: Please redo the changes in this file.

**Review by miracvbasaran — CHANGES_REQUESTED** [review]

**Review by miracvbasaran — APPROVED** [review]
